Sparse Merkle Tree and helpers

fubuloubu opened the pull request:
### What was wrong?
I wanted #58 to have more features that are useful for things like Plasma, like the ability to return Merkle branches, the ability to set key size/tree height, and to return node updates from setting a key so that you could use them to broadcast updates to other clients (see: [merkle sync](https://github.com/GunClear/merkle-sync/))

Solves #65

#### Cute Animal Picture
![baby walrus](http://blogs.discovermagazine.com/inkfish/files/2018/01/Screen-Shot-2018-01-12-at-1.35.26-PM.png)

fubuloubu committed:
Moved the 'sanity checks' to their own testcase

fubuloubu committed:
Added smt class and some basic testing

height -> depth; cleaner keysize check

Forgot height->depth in testcase

Added consistency check with branch and root

Moved calc_root to smt file, better as utility function for SMTs

pipermerriam commented on trie/smt.py:
```diff
@@ -0,0 +1,177 @@
+
+    def __init__(self, keysize=32):
+        # Ensure we can support the given height
+        if 1 > keysize or keysize > 32:
```
Maybe more legible:

```
if not 0 <= keysize <= 32:
   ...
```

pipermerriam commented on trie/smt.py:
```diff
@@ -0,0 +1,177 @@
+            raise ValidationError("Keysize must be number of bytes in range [0, 32]")
+
+        self.keysize = keysize  # keysize in bytes
+        self.height = keysize * 8  # height is number of bits in key
```
Is `height` a term from the spec.  I think `depth` might be another term we use for this measurement and for me at least it's a more natural name.  I don't feel strongly, but I'd default to pushing for `depth` absent an objective reason to use something else.

fubuloubu replied:
`depth` :+1:

pipermerriam commented on trie/smt.py:
```diff
@@ -0,0 +1,177 @@
+        self.height = keysize * 8  # height is number of bits in key
+        
+        # Initialize an empty tree with one branch
+        self.db = {}
```
I'd suggest following the same pattern as `HexaryTrie` and having the `db` be something that is passed in (so we can use non-memory based databases).

fubuloubu replied:
I built a class method for that purpose, since you have to initialize the database to a particular depth, with the root hash at the top. I could have that in the constructor, but I wanted to keep it separate to make it clearer that there is the potential for inconsistency if it accepts an outside db.

pipermerriam reviewed (commented):
Just some comments, didn't go through any of the core methods yet.

fubuloubu committed:
Added branch merging; finalized tests

Added branch merging functionality and tests, copied from merkle-sync

Moved to hypothesis strategies

Flake8

Added ability to set a default value, which does not fail existence check

Algorithm was complicated and had incorrect results; changed to clearer algorithm and eliminated bug

Updated name from nodes->node_updates to reflect how they are used; Added some comments

Turned branch merging into a method of a merkle proof manager class for SMTs

Linting issues

Added test case that would have detected borrowing error

Re-aligned the test case to capture the failure

lint error

Needed to make a deep copy to avoid error when using multiple Proof objects

Updates per snakecharmers' style guide

Type hints, test updates, a bug fix

- Added type hints
- Made some things that were lists into tuples
- Added some extra validations on types and sizes
- Needed to update some test cases for that
- Found a bug in SparseMerkleProof.merge()

Added docstrings for calc_root function

Added docstrings for SparseMerkleProof class

Commented out line with issue

removed commented out line stemming from older hypothesis in production

Updated to use more efficient data type

Forgot to remove unused import

pipermerriam commented on trie/smt.py:
```diff
@@ -0,0 +1,266 @@
+        for i in range(self.depth):
+            node_hash = keccak(node)
+            self.db[node_hash] = node
+            node = node_hash + node_hash
```
Experimented with how to do this without overriding the local `node` variable on each iteration.

```
    def do_iteration(value):
        node_hash = keccak(value)
        db[node_hash] = value
        return node_hash + node_hash

    root_node = pipe(
        node,
        *itertools.repeat(do_iteration, depth),
    )
    ...
```
    
Readability is reasonably poor so not a clear winner.

fubuloubu replied:
So... should I do anything to make this better?

pipermerriam replied:
> So... should I do anything to make this better?

I don't think so.  Most of the time when I see mutation I like to experiment with what it takes to remove it.  Often it isn't worth it (like this case) but I still like dumping the answer in case it's educational to anyone for how to do certain things using an functional approach.

pipermerriam commented on trie/smt.py:
```diff
@@ -0,0 +1,266 @@
+        self.db[self.root_hash] = node
+
+        # updates need to be in root->leaf order, so flip back
+        return list(reversed(proof_update))
```
We've tended to use `tuple` unless we have an explicit reason to need mutability.

fubuloubu replied:
Will make this update

pipermerriam commented on trie/smt.py:
```diff
@@ -0,0 +1,266 @@
+        node_hash = self.root_hash
+        # Append the sibling to the branch
+        # Iterate on the parent
+        for i in range(self.depth):
```
We have tended to use `_` for loop variables that we don't actually make use of.

fubuloubu replied:
Will do

pipermerriam commented on trie/smt.py:
```diff
@@ -0,0 +1,266 @@
+        self._key = key
+        self._keysize = len(key)
+        self._value = value
+        validate_length(branch, self._keysize*8)
```
Seems like it might be better to group both of the validation statements together at the beginning of the function body.

fubuloubu replied:
Does this mean I should pre-compute the number of bits the key contains?

pipermerriam replied:
No, literally just moving this statement up a few lines prior to setting the various values on `self`

pipermerriam commented on trie/smt.py:
```diff
@@ -0,0 +1,266 @@
+        self._value = value
+        validate_length(branch, self._keysize*8)
+        self._branch = deepcopy(branch)
+        self._branchsize = len(branch)
```
naming convention would be to have this as `_branch_size`

fubuloubu replied:
Will do

pipermerriam commented on trie/smt.py:
```diff
@@ -0,0 +1,266 @@
+    def __init__(self, key, value, branch):
+        validate_is_bytes(key)
+        self._key = key
+        self._keysize = len(key)
```
naming convention would have this be `_key_size`

fubuloubu replied:
Will do

pipermerriam reviewed (commented):
I'm still sort of struggling to understand how this would be used.  Maybe I missed it somewhere.  Are there code examples that might be illustrative here?

fubuloubu commented on tests/test_smt.py:
```diff
@@ -0,0 +1,79 @@
+    data=st.data(),
+    key_size=st.integers(min_value=1, max_value=32),
+)
+@settings(deadline=None)
```
This line does not work in hypothesis 3.7.0

fubuloubu commented on trie/smt.py:
```diff
@@ -0,0 +1,355 @@
+        >>> our_value = b'\x01'
+        >>> smt.set(our_key, our_value)
+        >>> # We need to track proof data for *some* reason
+        >>> our_proof = SparseMerkleProof(our_key, our_value, smt.branch(our_key))
```
Added a doctest to show off how this would be used

pipermerriam commented on tests/test_constants.py:
```diff
@@ -0,0 +1,39 @@
+    # More hashes the lower you go down the tree (to the root)
+    # NOTE: Did this with different code as a sanity check
+    for _ in range(DEPTH - 1):
+        EMPTY_NODE_HASHES.insert(0, keccak(EMPTY_NODE_HASHES[0] + EMPTY_NODE_HASHES[0]))
```
FYI, `list.insert(0, thing)` is one of the slowest standard library operations around.  My recommendation would be to either:

- build the list in reverse order and then reverse it once you are done
- use the `toolz.accumulate` function which does exactly what you need here
- use a `collections.deque` which has fast poping and insertion at both ends.

fubuloubu replied:
I will do this update, and try to squash some of it down.

fubuloubu replied:
Used `collections.deque`, but didn't really make a difference

pipermerriam commented on tests/test_smt.py:
```diff
@@ -0,0 +1,79 @@
+    data=st.data(),
+    key_size=st.integers(min_value=1, max_value=32),
+)
+#@settings(deadline=None)  # This option is not available in hypothesis-3.7.0
```
This like should just be removed.

fubuloubu replied:
It found a pretty decent bug I wouldn't have found otherwise, but can remove if upgrading isn't worth hassle.

pipermerriam replied:
My comment about removing it is that it commenting it out is equivalent to removing it but with the negative downside of leaving it in the codebase.  If we're leaving it in it should get a comment explaining why.  If we are no longer using hypothesis versions that support this option I don't see why we should leave this in...

carver commented on tests/test_smt.py:
```diff
@@ -0,0 +1,77 @@
+
+    # If you delete it, it goes away
+    smt.delete(k)
+    assert not smt.exists(k)
```
Maybe also confirm that the root hash is the same as it started?

carver commented on tests/test_constants.py:
```diff
@@ -0,0 +1,41 @@
+    # NOTE: Did this with different code as a sanity check
+    for _ in range(DEPTH - 1):
+        EMPTY_NODE_HASHES.appendleft(keccak(EMPTY_NODE_HASHES[0] + EMPTY_NODE_HASHES[0]))
+    EMPTY_ROOT_HASH = keccak(EMPTY_NODE_HASHES[0] + EMPTY_NODE_HASHES[0])
```
I think you only need to calculate this 32byte empty root hash once, and then save it as a constant to check against in the test (probably useful to have around anyhow). Putting the code in a utility for later verification/re-generation seems sufficient.

fubuloubu replied:
So, the root would change with different choices of the default. I'm not sure a utility would be a good suggestion, but I'm open to that. I wanted to calculate it here in a different way so that that test does not depend on the implementation. To put the branch explicitly would be almost 260 hardcoded lines, so I though this was a good compromise to doing that.

Also, we have the `smt.calc_root(k, v, b)` utility that can obtain a root hash.

carver commented on tests/test_smt.py:
```diff
@@ -0,0 +1,77 @@
+            unique=True,
+        )
+    )
+    vals = data.draw(
```
I haven't used data.draw before, but it seems like it makes it more difficult to reproduce the scenario if the parameters are generated at runtime instead of before the test. Doing it this other way also makes me wonder: is there a reason the standard way wouldn't work? (in this case: adding `num_elements`, `keys`, and `vals` as more parameters to `@given`)

fubuloubu replied:
I couldn't figure out how to make it work where the size of the keys in `keys` depended on the choice of `key_size` pulled from that strategy. I also needed the size of `vals` to match the size of `keys`, which also must be less than the number of possible options for the given `key_size`, although I suppose `unique=True` also guarantees this...

Open to suggestions on how to do that.

pipermerriam replied:
https://hypothesis.readthedocs.io/en/latest/data.html#hypothesis.strategies.shared

I think this is what you want.

```python
@given(
    key_size=st.shared(st.integers(min_value=1, max_value=32), key="key_size"),
    kv_pairs=st.shared(st.integers(min_value=1, max_value=32), key="key_size").flatmap(
       lambda key_size: st.tuples(st.binary(min_size=key_size, max_size=key_size), st.binary(min_size=1, max_size=1))
    ),
)
def test_thing(key_size, kv_pairs):
    ...
```
)

pipermerriam replied:
^ the above is psuedo-code but should get you going in the right direction.

fubuloubu replied:
Ahhhh, that's what the hell `st.shared` does!

Was trying to figure this out on the plane lol

fubuloubu replied:
Figured this out, thanks @pipermerriam!

carver commented on tests/test_smt.py:
```diff
@@ -0,0 +1,77 @@
+    # Track the big list of all updates
+    proof_updates = []
+    for k, p in proofs.items():
+        # Update the key in the smt a bunch of times
```
Why a bunch of times here? Why specifically are we doing this the number if times that there are keys? Is that a coincidence? I'm not clear why there are multiple resets of the values on the same key beyond testing one or two.

fubuloubu replied:
Yeah, I could see that argument. I wanted to show that even with sufficient amount of updates there is no loss of correctness. But this could occur with even 3 selections. Combined with the above issue with strategies, I could probably reduce this so that it's doing less unnecessary verification by having a hard coded number of updates.

fubuloubu replied:
Partially made this update by reducing the size of the list to a hard 3. Keeping the loop, because that means the update gets checked before, during, and after an update from another key. Tests are a little faster now :)

carver commented on tests/test_smt.py:
```diff
@@ -0,0 +1,77 @@
+
+        # Merge all of the updates into the tracked proof entries
+        for u in proof_updates:
+            p.merge(*u)
```
Generating the proof requires re-applying the relevant keys and values and the result of `smt.set()`?

fubuloubu replied:
`smt.set()` returns what the node updates are for a given change in value in the tree. These updates can be synchronized with a separate proof object (perhaps on a different computer), and that proof object will find the relevant sibling to keep itself up to date so that the proof object can track the same smt root separately without the extra data overhead of managing the whole tree (and the stream of all key-value updates). The storage size of managing the tree can grow non-linearly, whereas the proof object size will remain constant. This is at the expense of increased networking overhead, but that can be optimized by only communicating these updates to the practical depth of the tree (beyond which all keys map to single-branch sub-trees).

My intended use case is mobile, so this tradeoff of storage size is desirable. If my use case had 1m users, the tree would grow to be 2+ gigs in size (without any optimizations).

carver replied:
Sorry, this was a comment from before I read the whole thing. Meant to delete it, but it snuck in on my way out the door. Totally on board. :+1:

carver commented on trie/smt.py:
```diff
@@ -0,0 +1,355 @@
+)
+
+
+def calc_root(key: bytes, value: bytes, branch: Sequence[Hash32]) -> Hash32:
```
Maybe `siblings` instead of `branch`?

fubuloubu replied:
branch is a technical term, even the binary trie uses it: https://github.com/ethereum/py-trie/blob/master/trie/binary.py#L254

carver replied:
That's a branch *node*, to distinguish it from an extension or leaf node. (Perhaps I would have suggested `branching_node` instead, but not worth rehashing at this point)

A branch in a tree is the list of ancestors from the leaf to the root. The argument in this function is the list of the siblings of the ancestors.

One confirmation that it's a good name is that a loop below would change from:

```py
for sibling in branch:
```
to:

```py
for sibling in siblings:
```

fubuloubu replied:
changed all mentions of `sibling` to `sibling_node`. `branch` is still used to refer to an heirarchically-ordered set of a sibling nodes

carver commented on trie/smt.py:
```diff
@@ -0,0 +1,355 @@
+
+    :return: the root hash of the merkle proof computation
+
+    .. doctest::
```
Neat, is the repo set up to run doctests?

fubuloubu replied:
Nope lol.

I could turn this into an actual test case

fubuloubu replied:
So, these examples aren't too substantially different from the test cases to warrant creating more. I could remove it from the docstring to prevent an out-of-date example from weaseling it's way in.

carver commented on trie/smt.py:
```diff
@@ -0,0 +1,355 @@
+
+class SparseMerkleTree:
+
+    def __init__(self, key_size: int=32, default: bytes=BLANK_NODE):
```
Worth mentioning in a docstring that `key_size` is in bytes.

carver commented on trie/smt.py:
```diff
@@ -0,0 +1,355 @@
+    Track the current value and merkle proof branch for a given key.
+    This will enable the tracked proof data to stay up to date with changes
+    that may be streamed to the end user over external protocols without having
+    to interactively query the full SMT to obtain the most up-to-date branch.
```
I can see the value in not *having* to query the full SMT, but it seems like an option that should be supported. (I think it's fine to add an issue as a TODO, so this PR doesn't drag on).

It's worth noting that if this were used to generate proofs locally that it ~doubles the amount of hashing done, which is usually a significant computation in aggregate.

fubuloubu replied:
Not sure what you mean

fubuloubu replied:
Not sure what this comment was about, but noted that this could be handled in another PR if further issues are found with the approach.

carver commented on trie/smt.py:
```diff
@@ -0,0 +1,355 @@
+        >>> # tree updates can be communicated over any channel to proof obj
+        >>> our_proof.merge(their_key, their_new_value, their_node_updates)
+        >>> # Note our branch data was never queried from smt to our proof obj
+        >>> our_proof.branch == smt.branch(our_key)
```
I think `siblings` could be a good name for both this attribute and method, too.

fubuloubu replied:
Branch has a particular technical meaning though. "Siblings" isn't entirely accurate because they are at different levels in the tree.

Could brainstorm better words.

carver replied:
> Branch has a particular technical meaning though.

Okay, maybe I'll take back my definition of branch from above, I couldn't find any supporting sources. (Although I am not sure which technical meaning you're referring to either)

> "Siblings" isn't entirely accurate because they are at different levels in the tree.

Sure, in my head it's short for "siblings for the key nodes", or more pedantically "siblings for the nodes corresponding to each bit in the key". I still don't think it's a terrible name, but I'm happy to keep brainstorming.

Other options
- `uncles`
- `ancestor_uncles`
- `ancestor_siblings`
- `pruned_siblings`
- `proof_siblings`
- `proof_nodes`
- `key_siblings`

One reason why I still include `siblings`: the docstring and for loop both use sibling to refer to an individual node in the list.

fubuloubu replied:
`proof` might actually be most accurate. it also shows the use case a little clearer.

fubuloubu replied:
As mentioned above, changed all mentions of sibling to sibling_node. branch is still used to refer to an heirarchically-ordered set of a sibling nodes

carver commented on trie/smt.py:
```diff
@@ -0,0 +1,355 @@
+        >>> their_new_value = b'\x01'
+        >>> their_node_updates = smt.set(their_key, their_new_value)
+        >>> # tree updates can be communicated over any channel to proof obj
+        >>> our_proof.merge(their_key, their_new_value, their_node_updates)
```
`merge` to me usually means combining two things of equal type. Maybe `update()`? (like `dict`'s)

fubuloubu replied:
I mean, they are sort of equal type, but also not really.

`update` is good enough haha

carver commented on trie/smt.py:
```diff
@@ -0,0 +1,355 @@
+        self._key = key
+        self._key_size = len(key)
+        self._value = value
+        self._branch = list(copy.deepcopy(branch))  # Avoid issues with mutable lists
```
`Hash32` isn't mutable, so isn't `self._branch = list(siblings)` sufficient?

fubuloubu replied:
Fair. Reduces dependencies

carver commented on trie/smt.py:
```diff
@@ -0,0 +1,355 @@
+
+    @property
+    def root_hash(self) -> Hash32:
+        return calc_root(self.key, self.value, self.branch)
```
Probably want this in a cache, but I'm happy if that just goes in a follow-up issue.

fubuloubu replied:
You would probably check this only once before using it, which would update the cached value.

Not sure if a caching strategy would make sense, but open to seeing how the use cases play out.

carver replied:
Yup, totally happy to just punt :football:

carver commented on trie/smt.py:
```diff
@@ -0,0 +1,355 @@
+            #       Be sure to convert between them effectively.
+            for bit in reversed(range(self._branch_size)):
+                if path_diff & (1 << bit) > 0:
+                    branch_point = (self._branch_size-1)-bit
```
Yikes, how is `flake8` not complaining? It seems like it's running, but it should have required spaces around `-`.

carver commented on trie/smt.py:
```diff
@@ -0,0 +1,355 @@
+            #       to obtain the update. This allows an optimization
+            #       of pruning updates to the maximum possible depth
+            #       that would be required to merge, which may be many
+            #       orders of magnitude smaller than the tree depth.
```
"many orders of magnitude" sounds like a stretch. 256 values is only ~2 orders of magnitude larger than 1. :) "significantly" seems a sufficient word here.

fubuloubu replied:
Lol fair enough :)

carver commented on trie/smt.py:
```diff
@@ -0,0 +1,355 @@
+    def __init__(self, key_size: int=32, default: bytes=BLANK_NODE):
+        # Ensure we can support the given depth
+        if not 1 <= key_size <= 32:
+            raise ValidationError("Keysize must be number of bytes in range [0, 32]")
```
Why are we limited to 32? Not really important, but I don't see anything that prevents longer keys, yet.

fubuloubu replied:
That is true, but anything above 256 bit depth is impractical IMO. If it was suggested for a use case, I would counter by saying "hash the key"

carver replied:
Ok, maybe a comment would help, saying something like "the 32-byte limit is to purely encourage good performance characteristics"

carver commented on trie/smt.py:
```diff
@@ -0,0 +1,355 @@
+
+        self._default = default
+
+        # Initialize an empty tree with one branch
```
Ok, this time it can be a branch. :)

carver commented on trie/smt.py:
```diff
@@ -0,0 +1,355 @@
+
+        # If db is provided, and is not consistent,
+        # there may be a silent error. Can't solve that easily.
+        for key, value in db.items():
```
I don't think we can assume that the database is iterable. (or we often wouldn't want to, even if it is possible)

I suspect this can be modified to keep a reference to the database, and look up all the values on demand (like the other tries do).

fubuloubu replied:
Can you point to an example?

Also, what do you think about this being a separate class method?

carver replied:
> Can you point to an example?

Say, all of the state in Ethereum mainnet. Even though plyvel technically supports key iteration it would be prohibitively expensive ~~computationally~~ in I/O.

> Also, what do you think about this being a separate class method?

I think if this becomes like the other tries, with a reference to the underlying db, then most of this method can disappear, and the rest moves to init.

fubuloubu replied:
Well, unlike the Ethereum tries, there are two characteristics that I think warrant the separate approach.
1. SMTs can be used in a more ad-hoc fashion.  For example in Plasma, each transaction in a block is bundled into an SMT and summarized via the root, which is uploaded to Ethereum. Probably not too different from the txn root in Ethereum, but wortth noting.
2. You may want to specify a different default value for an empty database to the trie, so that it can be used for different characteristics. For example, storage in Ethereum assumes a default value for the key, which is `bytes32(0)` instead of `b''`. I was trying to allow for this behavior, and disambiguate it from db storage use cases through the separate class method.

carver replied:
On 1) nothing stops you from passing in a {} as the "underlying database", that's a totally reasonable use case that py-evm does sometimes, too.

On 2) if we want to be careful about it, `__init__` could always try to add the default values to the DB (maybe checking if they exist first). Other than that, it seems like we could still follow the same pattern for init in the other tries.

fubuloubu replied:
Re: 1) {} would not be a valid database, because it does not have a pathway to the default value.
Re: 2) because of 1), not sure if mixing this logic is the best choice because that would have an eiher/or constructor layout

carver replied:
1) huh? `self.db = {}` already in [the PR](https://github.com/ethereum/py-trie/pull/77/files/4f2e5f5e66f553b96bda5b76281a12309b2f16b3#diff-730d445fc351b9302eb71dc655598097R205)
2) I'm also cool with having a separate `smt = SMT.init_db(db, default_value=...)`, where `SMT(db, default_value=...)` would assume the database is already initialized.

fubuloubu replied:
1. Immediately followed by creating the branch for the default value in the db through the iteration lol
2. Cool, thanks. I think this is clearer that there are two use cases.

fubuloubu replied:
Solved the original concern by just directly assigning the db in the classmethod.

carver commented on trie/smt.py:
```diff
@@ -0,0 +1,355 @@
+        if value == BLANK_NODE:
+            raise KeyError("Key does not exist")
+
+        return value
```
Just a nicety if you like. We've found some success in doing `if/elif` tries even when they aren't *required* by control flow, like:

```py
        if value == BLANK_NODE:
            raise KeyError("Key does not exist")
        else:
            return value
```

fubuloubu replied:
Does that lead to an optimization or better code analysis?

carver replied:
It's all about readability: the visual confirmation that this is a branching situation and the lines below it only happen in the alternate case.

davesque replied:
@carver Maybe too many cooks here, but isn't it also common for guarding (i.e. stack popping) if statements to lack else clauses?  I know that I often write code in this pattern and sort of prefer it since it's not hard to tell at a glance that the body of the if statement would prevent any further execution in that subroutine.

carver replied:
Yeah, my personal preference is to only do those at the very beginning of a method, rather than after executing some logic. But this is just a light preference, definitely not a required change. `#WONTFIX` is fine

carver commented on trie/smt.py:
```diff
@@ -0,0 +1,355 @@
+                node_hash = self.db[node_hash][32:]
+            else:
+                branch.append(self.db[node_hash][32:])
+                node_hash = self.db[node_hash][:32]
```
Maybe you would like:

```py
        for _ in range(self.depth):
            node = self.db[node_hash]
            left, right = node[:32], node[32:]

            if path & target_bit:
                node_hash = right
                siblings.append(left)
            else:
                node_hash = left
                siblings.append(right)

fubuloubu replied:
Yeah, that's clearer.

carver reviewed (changes requested):
Ok, *almost* got through it before I had to take off. No huge problems, but I'll add an X since Piper already gave it a :heavy_check_mark: , as a reminder that there are still a few things to go over.

I really think you're going to want to do the db like the other trie's do, to make it easy to have a disk-backed db.

Also, you'll probably want something like `squash_changes` so that you don't flood the db with a bunch of intermediate state that you don't care about in the middle of a transaction:
https://github.com/ethereum/py-trie/blob/bce9e027b0ef4f411c2eabc0600e6b94679075d8/trie/hexary.py#L444-L450

fubuloubu committed:
Added updates per PR review

carver commented on trie/smt.py:
```diff
@@ -0,0 +1,355 @@
+        node = self._default  # Default leaf node
+        for _ in range(self.depth):
+            node_hash = keccak(node)
+            self.db[node_hash] = node
```
One optimization (probably to go in an issue for later): it looks like these node hashes could be stored in a blank hash ancestor lookup so the `get()` can shortcut and return a blank result as soon as it hits one of the blank hash ancestors. (making it `O(log(h))` instead of `O(h)` operations to find a blank key)

fubuloubu replied:
I think we can leave this out for now, but definitely something to think about.

fubuloubu committed:
Added a comment about keysize

fubuloubu committed:
Better way of ensuring the key_size maps to the size of the keys chosen

fubuloubu committed:
Lint issues

carver commented on trie/smt.py:
```diff
@@ -0,0 +1,369 @@
+
+        # If db is provided, and is not consistent,
+        # there may be a silent error. Can't solve that easily.
+        smt.db = db
```
Reminder that we probably want the ability to initialize a passed-in database with empty node hashes (eg~ LevelDB).
